rclcpp_lifecycle/src/lifecycle_node_interface_impl.hpp

```diff
@@ -0,0 +1,401 @@
+      fprintf(stderr, "%s:%d, Unable to change state for state machine for %s: %s \n",
+        __FILE__, __LINE__, node_base_interface_->get_name(), rcl_get_error_string_safe());
+      return false;
+    }
+
+    // keep the initial state to pass to a transition callback
+    State initial_state(state_machine_.current_state);
+
+    unsigned int transition_id = static_cast<unsigned int>(lifecycle_transition);
+    if (rcl_lifecycle_trigger_transition(&state_machine_, transition_id, RCL_RET_OK,
+      true) != RCL_RET_OK)
+    {
+      fprintf(stderr, "%s:%d, Unable to start transition %u from current state %s: %s\n",
+        __FILE__, __LINE__, transition_id,
+        state_machine_.current_state->label, rcl_get_error_string_safe());
+      return false;
+    }
+
+    auto cb_success = execute_callback(
+      state_machine_.current_state->id, initial_state);
+
+    if (rcl_lifecycle_trigger_transition(
+        &state_machine_, transition_id, cb_success, true) != RCL_RET_OK)
+    {
+      fprintf(stderr, "Failed to finish transition %u. Current state is now: %s\n",
+        transition_id, state_machine_.current_state->label);
+      return false;
+    }
+
+    // error handling ?!
+    // TODO(karsten1987): iterate over possible ret value
+    if (cb_success == RCL_LIFECYCLE_RET_ERROR) {
+      auto error_resolved = execute_callback(state_machine_.current_state->id, initial_state);
+      if (error_resolved == RCL_RET_OK) {
+        // We call cleanup on the error state
+        rcl_lifecycle_trigger_transition(
+          &state_machine_, lifecycle_msgs__msg__Transition__TRANSITION_CLEANUP, true, true);
+      } else {
+        // We call shutdown on the error state
+        rcl_lifecycle_trigger_transition(
+          &state_machine_, lifecycle_msgs__msg__Transition__TRANSITION_SHUTDOWN, true, true);
+      }
+    }
+    // This true holds in both cases where the actual callback
+    // was successful or not, since at this point we have a valid transistion
+    // to either a new primary state or error state
+    return true;
+  }
+
+  rcl_lifecycle_ret_t
+  execute_callback(unsigned int cb_id, const State & previous_state)
+  {
+    // in case no callback was attached, we forward directly
+    auto cb_success = RCL_LIFECYCLE_RET_OK;
+
+    auto it = cb_map_.find(cb_id);
+    if (it != cb_map_.end()) {
+      auto callback = it->second;
+      try {
+        cb_success = callback(State(previous_state));
+      } catch (const std::exception & e) {
+        fprintf(stderr, "Caught exception in callback for transition %d\n",
+          it->first);
+        fprintf(stderr, "Original error msg: %s\n", e.what());
+        // maybe directly go for error handling here
+        // and pass exception along with it
+        cb_success = RCL_LIFECYCLE_RET_ERROR;
+      }
+    } else {
+      fprintf(stderr, "%s:%d, No callback is registered for transition %u.\n",
+        __FILE__, __LINE__, cb_id);
```

Comment:
Is it appropriate to always print a (warning?) message to stderr when no callback is associated?